pandas/io/common.py

```diff
@@ -192,13 +219,7 @@ def get_filepath_or_buffer(filepath_or_buffer, encoding=None,
     filepath_or_buffer = _stringify_path(filepath_or_buffer)
 
     if _is_url(filepath_or_buffer):
-        req = _urlopen(filepath_or_buffer)
-        content_encoding = req.headers.get('Content-Encoding', None)
-        if content_encoding == 'gzip':
-            # Override compression based on Content-Encoding header
-            compression = 'gzip'
-        reader = BytesIO(req.read())
-        req.close()
+        reader, compression = _urlopen(filepath_or_buffer, session=session)
```

Comment:
`compression` is not mentioned in the docs and I wonder if `session` should.
If so we can use the template/Appender approach, just not sure how to do that in a clean way.

Reply:
1) `compression` should be mentioned in the docs.  Would be great to add that.
2) Given my first point, `session` should be as well.

And yes, template / Appender approach sounds like a good idea!

Reply:
I need some guidance on how to do this when crossing modules, `common.py` in `get_filepath_or_buffer` and `parsers.py` for all the parsers defined in `_parser_params`.

The options are:

| options               | `get_filepath_or_buffer`    | `_parser_params`  |
|-----------------------|-----------------------------|-------------------|
| `filepath_or_buffer`  |  present                    | NA                |
| `mode`                |  present                    | NA                |
| `encoding`            |  present                    | present           |
| `compression`         |  missing (added in this PR) | present           |
| `session`             |  missing                    | present           |

It is unclear to me how to fix this. Some questions I have:
- Should the `compression` and `encoding` docstring be standardized?
- If yes on the question above what is the best strategy for this? Should `encoding`, `compression`, and `session` be defined in `common.py` and imported in `parsers.py` for appending in `_parser_params`? Or the other way around?

My guess is that I should do get the common options from `parser.py` and add it to `common.py` like:

```python

_common_params = r"""
Parameters
----------
encoding : str, default None
    Encoding to use for UTF when reading/writing (ex. 'utf-8'). `List of Python
    standard encodings
    <https://docs.python.org/3/library/codecs.html#standard-encodings>`_
session : requests.Session
    object with the a requests session configuration for remote file.
    (requires the requests library)
compression : {'infer', 'gzip', 'bz2', 'zip', 'xz', None}, default 'infer'
    For on-the-fly decompression of on-disk data. If 'infer' and
    `filepath_or_buffer` is path-like, then detect compression from the
    following extensions: '.gz', '.bz2', '.zip', or '.xz' (otherwise no
    decompression). If using 'zip', the ZIP file must contain only one data
    file to be read in. Set to None for no decompression.

    .. versionadded:: 0.18.1 support for 'zip' and 'xz' compression.
"""
```

Then compose the `get_filepath_or_buffer` with the missing `filepath_or_buffer` and `mode`, and later `import` `_common_params` in `parsers.py` to compose `_parser_params`. Does that sound OK?

PS: if that is correct I'd prefer if a pandas doc expert did this instead of me, I'm kind of lost in the `docstrings` and `Appender`s :smile: